File: src/main/java/org/flossware/civilization/util/ScenarioBuilder.java
package org.flossware.civilization.util;

import org.flossware.civilization.model.*;

import java.util.*;

/**
 * Builder for constructing scenarios programmatically.
 */
public final class ScenarioBuilder {
    private String scenarioId = "scenario-" + UUID.randomUUID();
    private String name = "Unnamed Scenario";
    private String description = "";
    private int startYear = -3000;
    private int endYear = 2026;

    private CivilizationState initialState;
    private final List<Technology> techTree = new ArrayList<>();
    private WorldConstraints worldConstraints;
    private SimulationRules simulationRules;

    public ScenarioBuilder withId(String id) {
        this.scenarioId = id;
        return this;
    }

    public ScenarioBuilder withName(String name) {
        this.name = name;
        return this;
    }

    public ScenarioBuilder withDescription(String description) {
        this.description = description;
        return this;
    }

    public ScenarioBuilder withTimeRange(int startYear, int endYear) {
        this.startYear = startYear;
        this.endYear = endYear;
        return this;
    }

    public ScenarioBuilder withInitialState(CivilizationState state) {
        this.initialState = state;
        return this;
    }

    public ScenarioBuilder addTechnology(Technology tech) {
        this.techTree.add(tech);
        return this;
    }

    public ScenarioBuilder withTechTree(List<Technology> techs) {
        this.techTree.clear();
        this.techTree.addAll(techs);
        return this;
    }

    public ScenarioBuilder withWorldConstraints(WorldConstraints constraints) {
        this.worldConstraints = constraints;
        return this;
    }

    public ScenarioBuilder withSimulationRules(SimulationRules rules) {
        this.simulationRules = rules;
        return this;
    }

    public Scenario build() {
        Objects.requireNonNull(initialState, "Initial state must be set");
        Objects.requireNonNull(worldConstraints, "World constraints must be set");
        Objects.requireNonNull(simulationRules, "Simulation rules must be set");

        return new Scenario(
            scenarioId,
            name,
            description,
            startYear,
            endYear,
            initialState,
            techTree,
            worldConstraints,
            simulationRules
        );
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/util/ScenarioBuilder.kt
package org.flossware.civilization.util

import org.flossware.civilization.model.*

class ScenarioBuilder {
    private var scenarioId: String = "scenario-${kotlin.random.Random.nextLong()}"
    private var name: String = "Unnamed Scenario"
    private var description: String = ""
    private var startYear: Int = -3000
    private var endYear: Int = 2026
    private var initialState: CivilizationState? = null
    private val techTree: MutableList<Technology> = mutableListOf()
    private var worldConstraints: WorldConstraints? = null
    private var simulationRules: SimulationRules? = null

    fun withId(id: String) = apply { scenarioId = id }
    fun withName(name: String) = apply { this.name = name }
    fun withDescription(description: String) = apply { this.description = description }
    fun withTimeRange(startYear: Int, endYear: Int) = apply { this.startYear = startYear; this.endYear = endYear }
    fun withInitialState(state: CivilizationState) = apply { initialState = state }
    fun addTechnology(tech: Technology) = apply { techTree.add(tech) }
    fun withTechTree(techs: List<Technology>) = apply { techTree.clear(); techTree.addAll(techs) }
    fun withWorldConstraints(constraints: WorldConstraints) = apply { worldConstraints = constraints }
    fun withSimulationRules(rules: SimulationRules) = apply { simulationRules = rules }

    fun build(): Scenario {
        return Scenario(
            scenarioId = scenarioId,
            name = name,
            description = description,
            startYear = startYear,
            endYear = endYear,
            initialState = requireNotNull(initialState) { "Initial state must be set" },
            techTree = techTree.toList(),
            worldConstraints = requireNotNull(worldConstraints) { "World constraints must be set" },
            simulationRules = requireNotNull(simulationRules) { "Simulation rules must be set" }
        )
    }
}
